antelope_contracts/contracts/erc20/src/erc20.cpp

```diff
@@ -8,92 +8,269 @@
-    // Assume account opened in evm_runtime
+    bytes value_zero; 
+    value_zero.resize(32, 0);
+
+    // required account opened in evm_runtime
     call_action call_act(evm_account, {{get_self(), "active"_n}});
-    call_act.send(get_self(), to, 0, call_data, evm_init_gaslimit);
+    call_act.send(get_self(), to, value_zero, call_data, evm_init_gaslimit);
+
+
+    evmc::address impl_addr = silkworm::create_address(reserved_addr, my_next_nonce()); 
 
-    // Assume nonce...
-    auto deploy_addr = silkworm::create_address(reserved_addr, 0); 
+    impl_contract_table_t contract_table(_self, _self.value);
+    contract_table.emplace(_self, [&](auto &v) {
+        v.id = contract_table.available_primary_key();
+        v.address.resize(kAddressLength);
+        memcpy(&(v.address[0]), impl_addr.bytes, kAddressLength);
+    });
+}
+
+[[eosio::action]] void erc20::regtoken(eosio::name eos_contract_name, std::string evm_token_name, std::string evm_token_symbol, const eosio::asset& min_deposit, const eosio::asset& deposit_fee, const eosio::asset &egress_fee, std::string erc20_impl_address, int erc20_precision) {
+    require_auth(get_self());
 
+    eosio::check(eos_contract_name.value, "invalid eos_contract_name");
+    
+    // keep the name & symbol fit into 32 byte, which is the alignment in EVM
+    eosio::check(evm_token_name.length() > 0 && evm_token_name.length() < 32, "invalid evm_token_name length");
+    eosio::check(evm_token_symbol.length() > 0 && evm_token_symbol.length() < 32, "invalid evm_token_symbol length");
+
+    // 2^(256-64) = 6.2e+57, so the precision diff is at most 57
+    eosio::check(erc20_precision >= min_deposit.symbol.precision() &&
+    erc20_precision <= min_deposit.symbol.precision() + 57, "erc20 precision out of range");
+
+    std::optional<bytes> impl_address_bytes = from_hex(erc20_impl_address);
+    eosio::check(!!impl_address_bytes && impl_address_bytes->size() == kAddressLength, "invalid erc20 address");
+
+    eosio::check(min_deposit.symbol == deposit_fee.symbol, "deposit_fee should have the same symbol of min_deposit");
+
+    eosio::check(egress_fee.symbol == native_token_symbol, "egress_fee should have native token symbol");
+    intx::uint256 egress_fee_evm = egress_fee.amount;
+    egress_fee_evm *= minimum_natively_representable;
+
+    uint128_t v = eos_contract_name.value;
+    v <<= 64;
+    v |= min_deposit.symbol.code().raw();
+    token_table_t token_table(_self, _self.value);
+    auto index_symbol = token_table.get_index<"by.symbol"_n>();
+    check(index_symbol.find(v) == index_symbol.end(), "token already registered");
+
+    impl_contract_table_t contract_table(_self, _self.value);
+    auto index = contract_table.get_index<"by.address"_n>();
+    auto itr = index.find(make_key(*impl_address_bytes));
+
+    eosio::check(itr != index.end(), "implementation contract must be deployed via erc20::init()");
+
+    auto reserved_addr = silkworm::make_reserved_address(get_self().value);
```

Comment:
This `get_self()` would also instead be the variable holding the name for `eosio.erc2o`.